client/dvm/rpc/src/eth.rs

```diff
@@ -793,21 +868,63 @@ where
 		};
 
 		let chain_id = match self.chain_id() {
-			Ok(chain_id) => chain_id,
+			Ok(Some(chain_id)) => chain_id.as_u64(),
+			Ok(None) => return Box::pin(future::err(internal_err("chain id not available"))),
 			Err(e) => return Box::pin(future::err(e)),
 		};
 
-		let message = LegacyTransactionMessage {
-			nonce,
-			gas_price: request.gas_price.unwrap_or(U256::from(1u32)),
-			gas_limit: request.gas.unwrap_or(U256::max_value()),
-			value: request.value.unwrap_or(U256::zero()),
-			input: request.data.map(|s| s.into_vec()).unwrap_or_default(),
-			action: match request.to {
-				Some(to) => TransactionAction::Call(to),
-				None => TransactionAction::Create,
-			},
-			chain_id: chain_id.map(|s| s.as_u64()),
+		let hash = self.client.info().best_hash;
+
+		let gas_price = request.gas_price;
+		let gas_limit = match request.gas {
+			Some(gas_limit) => gas_limit,
+			None => {
+				let block = self
+					.client
+					.runtime_api()
+					.current_block(&BlockId::Hash(hash));
+				if let Ok(Some(block)) = block {
+					block.header.gas_limit
```

Comment:
Is there any contradiction between here with https://github.com/darwinia-network/darwinia-common/pull/1040/files#diff-9973ba6af00d525497b89265da62e5912765910b517f15f6e9717abfe480cf16R472?

Reply:
O, no contradiction. the gas_limit in this RPC will be set to block_limit only if the request's gas is absent. I only keep this the same with upstream. We will reuse the upstream eth client in the next pr.